File: src/scanner.py
import json
from datetime import datetime
from pathlib import Path
from typing import Any

from .toolchains import brew, npm, pip, gem, cargo, docker

REPORT_FILE = Path(__file__).resolve().parent.parent / "report.json"

SCANNERS = {
    "brew": brew.scan,
    "npm": npm.scan,
    "pip": pip.scan,
    "gem": gem.scan,
    "cargo": cargo.scan,
    "docker": docker.scan,
}

_LABELS = {"ok": "PASS", "warning": "WARN", "error": "FAIL", "skipped": "SKIP"}


def run_scan(chain: str = "all"):
    if chain == "all":
        results = {name: fn() for name, fn in SCANNERS.items()}
    elif chain in SCANNERS:
        results = {chain: SCANNERS[chain]()}
    else:
        return {
            "error": f"Unknown chain: {chain}. Options: all, {', '.join(SCANNERS.keys())}"
        }

    report = {
        "timestamp": datetime.now().isoformat(),
        "results": results,
    }

    REPORT_FILE.write_text(json.dumps(report, indent=2))
    return report


def _extract_outdated(res: dict) -> list[dict[str, Any]]:
    items = []
    for key in ("outdated_formulae", "outdated_casks", "outdated_global", "outdated"):
        if key in res and res[key]:
            for item in res[key]:
                if isinstance(item, dict) and "name" in item:
                    items.append(item)
    return items


def format_report(report: dict = None) -> str:
    if report is None:
        if REPORT_FILE.exists():
            report = json.loads(REPORT_FILE.read_text())
        else:
            return "No report available. Run `scan` first."

    results = report["results"]
    lines = []
    lines.append("# Envexa Health Report")
    lines.append(f"**Generated:** {report['timestamp']}")
    lines.append("")

    outdated_all: dict[str, list[dict[str, Any]]] = {}
    dashboard_rows = []

    for tool, res in results.items():
        outdated_items = _extract_outdated(res)
        if outdated_items:
            outdated_all[tool] = outdated_items

        status_text = {"ok": "PASS", "warning": f"WARN ({len(outdated_items)})", "error": "FAIL", "skipped": "SKIP"}.get(res["status"], "?")
        version_keys = {"version": "Version", "node_version": "Node", "python_version": "Python", "ruby_version": "Ruby", "rustc_version": "Rust"}
        ver_str = ""
        for k in version_keys:
            if k in res and res[k]:
                ver_str = res[k]
                break
        dashboard_rows.append(f"| {tool.title():8} | {status_text:<18} | {ver_str} |")

    lines.append("## Dashboard")
    lines.append(f"| {'Toolchain':8} | {'Status':<18} | {'Version'} |")
    lines.append(f"|{'-'*10}|{'-'*20}|{'-'*20}|")
    for row in dashboard_rows:
        lines.append(row)
    lines.append("")

    if outdated_all:
        lines.append("## Outdated Packages")
        lines.append("```")
        tool_names = list(outdated_all.keys())
        for i, tool in enumerate(tool_names):
            items = outdated_all[tool]
            is_last_tool = i == len(tool_names) - 1
            tool_prefix = "└── " if is_last_tool else "├── "
            lines.append(f"{tool_prefix}{tool.title()} ({len(items)})")
            indent_prefix = "    " if is_last_tool else "│   "
            for j, item in enumerate(items):
                sub_prefix = "└── " if j == len(items) - 1 else "├── "
                ver = f"{item.get('current', '?')} -> {item.get('latest', '?')}"
                lines.append(f"{indent_prefix}{sub_prefix}{item['name']}: {ver}")
        lines.append("```")
        lines.append("")

    lines.append("## Per-Toolchain Details")
    lines.append("")
    for tool, res in results.items():
        label = _LABELS.get(res["status"], "?")
        lines.append(f"### [{label}] {tool.title()}")

        if res["status"] == "skipped":
            lines.append(f"> {res['issues'][0] if res['issues'] else 'Skipped'}")
            lines.append("")
            continue

        version_labels = {"version": "Version", "node_version": "Node", "python_version": "Python", "ruby_version": "Ruby", "rustc_version": "Rust", "cargo_version": "Cargo"}
        ver_parts = []
        for k, v in version_labels.items():
            if k in res and res[k]:
                ver_parts.append(f"**{v}:** {res[k]}")
        if ver_parts:
            lines.append(" | ".join(ver_parts))

        installed = res.get("installed_count")
        if installed:
            lines.append(f"**Formulae:** {installed}")

        disk = res.get("disk_usage")
        if disk:
            for typ, info in disk.items():
                lines.append(f"- **{typ}:** {info.get('size', '?')} ({info.get('reclaimable', '?')} reclaimable)")

        outdated_items = _extract_outdated(res)
        if outdated_items:
            lines.append("")
            lines.append("| Package | Current | Latest |")
            lines.append("|---------|---------|--------|")
            for item in outdated_items:
                lines.append(f"| {item['name']} | {item.get('current', '?')} | {item.get('latest', '?')} |")

        tool_issues = [i for i in res.get("issues", []) if i]
        if tool_issues:
            lines.append("")
            for i in tool_issues:
                lines.append(f"> {i}")

        lines.append("")

    return "\n".join(lines)

File: src/scanner.rs
use std::collections::HashMap;

use crate::toolchains::{PackageInfo, ScanResult};

struct Table {
    headers: Vec<String>,
    rows: Vec<Vec<String>>,
}

impl Table {
    fn new() -> Self {
        Table {
            headers: Vec::new(),
            rows: Vec::new(),
        }
    }

    fn header(&mut self, cols: &[&str]) {
        self.headers = cols.iter().map(|c| c.to_string()).collect();
    }

    fn add_row(&mut self, cols: &[&str]) {
        self.rows.push(cols.iter().map(|c| c.to_string()).collect());
    }

    fn render(&self) -> String {
        let ncols = self.headers.len();
        if ncols == 0 {
            return String::new();
        }

        let mut widths = vec![0usize; ncols];
        for (i, h) in self.headers.iter().enumerate() {
            widths[i] = widths[i].max(h.len());
        }
        for row in &self.rows {
            for (i, cell) in row.iter().enumerate() {
                if i < ncols {
                    widths[i] = widths[i].max(cell.len());
                }
            }
        }

        let sep_body: Vec<String> = widths.iter().map(|w| "-".repeat(w + 2)).collect();
        let sep = format!("+{}+", sep_body.join("+"));

        let fmt = |cells: &[String]| -> String {
            let mut s = String::from("|");
            for (i, cell) in cells.iter().enumerate() {
                if let Some(&w) = widths.get(i) {
                    s.push_str(&format!(" {:<w$} |", cell, w = w));
                } else {
                    s.push_str(&format!(" {} |", cell));
                }
            }
            s
        };

        let mut out = String::new();
        out.push_str(&sep);
        out.push('\n');
        out.push_str(&fmt(&self.headers));
        out.push('\n');
        out.push_str(&sep);
        out.push('\n');
        for row in &self.rows {
            out.push_str(&fmt(row));
            out.push('\n');
        }
        out.push_str(&sep);
        out
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Report {
    pub timestamp: String,
    pub results: HashMap<String, ScanResult>,
}

const LABELS: &[(&str, &str)] = &[
    ("ok", "PASS"),
    ("warning", "WARN"),
    ("error", "FAIL"),
    ("skipped", "SKIP"),
];

const DISPLAY_NAMES: &[(&str, &str)] = &[
    ("brew", "Brew"),
    ("npm", "npm"),
    ("pnpm", "pnpm"),
    ("yarn", "Yarn"),
    ("bun", "Bun"),
    ("deno", "Deno"),
    ("pip", "pip"),
    ("gem", "Gem"),
    ("cargo", "Cargo"),
    ("docker", "Docker"),
];

fn tool_order() -> [&'static str; 10] {
    [
        "brew", "npm", "pnpm", "yarn", "bun", "deno", "pip", "gem", "cargo", "docker",
    ]
}

fn display_name(tool: &str) -> &str {
    DISPLAY_NAMES
        .iter()
        .find(|(k, _)| *k == tool)
        .map(|(_, v)| *v)
        .unwrap_or(tool)
}

fn status_label(s: &str) -> &str {
    LABELS
        .iter()
        .find(|(k, _)| *k == s)
        .map(|(_, v)| *v)
        .unwrap_or("?")
}

fn extract_outdated(res: &ScanResult) -> Vec<&PackageInfo> {
    let mut items = vec![];
    for key in [
        "outdated_formulae",
        "outdated_casks",
        "outdated_global",
        "outdated",
    ] {
        match key {
            "outdated_formulae" => items.extend(res.outdated_formulae.iter()),
            "outdated_casks" => items.extend(res.outdated_casks.iter()),
            "outdated_global" => items.extend(res.outdated_global.iter()),
            "outdated" => items.extend(res.outdated.iter()),
            _ => {}
        }
    }
    items
}

fn first_version(res: &ScanResult) -> String {
    let fields = [
        "version",
        "node_version",
        "python_version",
        "ruby_version",
        "rustc_version",
        "cargo_version",
        "pnpm_version",
        "bun_version",
        "deno_version",
    ];
    for f in fields {
        let val = match f {
            "version" => &res.version,
            "node_version" => &res.node_version,
            "python_version" => &res.python_version,
            "ruby_version" => &res.ruby_version,
            "rustc_version" => &res.rustc_version,
            "cargo_version" => &res.cargo_version,
            "pnpm_version" => &res.pnpm_version,
            "bun_version" => &res.bun_version,
            "deno_version" => &res.deno_version,
            _ => &None,
        };
        if let Some(v) = val {
            return v.clone();
        }
    }
    String::new()
}

pub fn format_report(report: &Report) -> String {
    let results = &report.results;
    let mut lines = vec![];
    lines.push("# Envexa Health Report".into());
    lines.push(format!("**Generated:** {}", report.timestamp));
    lines.push(String::new());

    let mut outdated_all: HashMap<&str, Vec<&PackageInfo>> = HashMap::new();

    for tool in &tool_order() {
        if let Some(res) = results.get(*tool) {
            let items = extract_outdated(res);
            if !items.is_empty() {
                outdated_all.insert(tool, items);
            }
        }
    }

    lines.push("## Dashboard".into());
    let mut dt = Table::new();
    dt.header(&["Toolchain", "Status", "Version"]);
    for tool in &tool_order() {
        if let Some(res) = results.get(*tool) {
            let display = display_name(tool);
            let label = status_label(&res.status);
            let ver = first_version(res);
            dt.add_row(&[display, label, &ver]);
        }
    }
    lines.push(dt.render());
    lines.push(String::new());

    if !outdated_all.is_empty() {
        lines.push("## Outdated Packages".into());
        let mut ot = Table::new();
        ot.header(&["Toolchain", "Package", "Current", "Latest"]);
        for tool in &tool_order() {
            if let Some(items) = outdated_all.get(tool) {
                let display = display_name(tool);
                for item in items {
                    ot.add_row(&[display, &item.name, &item.current, &item.latest]);
                }
            }
        }
        lines.push(ot.render());
        lines.push(String::new());
    }

    lines.push("## Per-Toolchain Details".into());
    lines.push(String::new());

    for tool in &tool_order() {
        if let Some(res) = results.get(*tool) {
            let label = status_label(&res.status);
            let display = display_name(tool);
            lines.push(format!("### [{label}] {display}"));

            if res.status == "skipped" {
                let reason = res.issues.first().map(|s| s.as_str()).unwrap_or("Skipped");
                lines.push(format!("> {reason}"));
                lines.push(String::new());
                continue;
            }

            let version_labels = [
                ("version", &res.version),
                ("node_version", &res.node_version),
                ("python_version", &res.python_version),
                ("ruby_version", &res.ruby_version),
                ("rustc_version", &res.rustc_version),
                ("cargo_version", &res.cargo_version),
                ("pnpm_version", &res.pnpm_version),
                ("bun_version", &res.bun_version),
                ("deno_version", &res.deno_version),
            ];
            let ver_parts: Vec<String> = version_labels
                .iter()
                .filter_map(|(key, val)| {
                    val.as_ref().map(|v| {
                        let label = match *key {
                            "version" => "Version",
                            "node_version" => "Node",
                            "python_version" => "Python",
                            "ruby_version" => "Ruby",
                            "rustc_version" => "Rust",
                            "cargo_version" => "Cargo",
                            "pnpm_version" => "pnpm",
                            "bun_version" => "Bun",
                            "deno_version" => "Deno",
                            _ => key,
                        };
                        format!("**{label}:** {v}")
                    })
                })
                .collect();
            if !ver_parts.is_empty() {
                lines.push(ver_parts.join(" | "));
            }

            if let Some(count) = res.installed_count {
                lines.push(format!("**Formulae:** {count}"));
            }

            if let Some(ref disk) = res.disk_usage {
                if let Some(obj) = disk.as_object() {
                    for (typ, info) in obj {
                        let size = info.get("size").and_then(|v| v.as_str()).unwrap_or("?");
                        lines.push(format!("- **{typ}:** {size}"));
                    }
                }
            }

            let outdated_items = extract_outdated(res);
            if !outdated_items.is_empty() {
                let mut pt = Table::new();
                pt.header(&["Package", "Current", "Latest"]);
                for item in &outdated_items {
                    pt.add_row(&[&item.name, &item.current, &item.latest]);
                }
                lines.push(pt.render());
            }

            for issue in &res.issues {
                lines.push(format!("> {issue}"));
            }

            lines.push(String::new());
        }
    }

    lines.join("\n")
}

pub fn format_status(report: &Report) -> String {
    let results = &report.results;
    let mut lines = vec![];
    lines.push("# Envexa Status".into());
    lines.push(format!("**Generated:** {}", report.timestamp));
    lines.push(String::new());

    let mut t = Table::new();
    t.header(&["Toolchain", "Status", "Count"]);
    for tool in &tool_order() {
        if let Some(res) = results.get(*tool) {
            let label = status_label(&res.status);
            let n = extract_outdated(res).len();
            let count = if n > 0 { n.to_string() } else { "-".into() };
            let display = display_name(tool);
            t.add_row(&[display, label, &count]);
        }
    }
    lines.push(t.render());

    lines.push(String::new());
    lines.push("Run `/envexa:scan` for full report or `/envexa:outdated` for details.".into());
    lines.join("\n")
}

pub fn format_outdated(report: &Report) -> String {
    let results = &report.results;

    let mut t = Table::new();
    t.header(&["Toolchain", "Package", "Current", "Latest"]);
    let mut has_anything = false;
    for tool in &tool_order() {
        if let Some(res) = results.get(*tool) {
            let items = extract_outdated(res);
            if !items.is_empty() {
                has_anything = true;
                let display = display_name(tool);
                for item in &items {
                    t.add_row(&[display, &item.name, &item.current, &item.latest]);
                }
            }
        }
    }

    if !has_anything {
        return "All packages are up to date!".into();
    }

    format!("# Outdated Packages\n\n{}", t.render())
}
